File: 283.py
# LeetCode 283 Move Zeroes
# Reference: https://www.youtube.com/watch?v=aayNRwUN3Do

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # QuickSort
        l = 0
        for r in range(len(nums)):
            if nums[r]: # nums[r] == 0
                nums[l], nums[r] = nums[r], nums[l]
                l += 1
        return nums
        
#  Time Complexity is O(N) and Space Complexity is O(1)

File: 283.java
//LeetCode 283 Move Zeroes
// Reference: https://www.youtube.com/watch?v=k5lIW5XxC7g

// Time Complexity is O(N) and Space Complexity is O(1)
class Solution{
    public void moveZeroes(int[] nums){
        // Start with the first position
        int insertPosition = 0;

        for (int i = 0; i < nums.length; i++){
            //Move all non-zero integers to left side
            if(nums[i] != 0){
                nums[insertPosition] = nums[i];
                insertPosition ++;
            }
        }

        while (insertPosition < nums.length){
            nums[insertPosition ++] = 0;
        }
    }

}

